packages/contracts-bedrock/src/dispute/weth/DelayedWETH.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { ISemver } from "src/universal/ISemver.sol";
+
+import { IDelayedWETH } from "src/dispute/interfaces/IDelayedWETH.sol";
+import { IWETH } from "src/dispute/interfaces/IWETH.sol";
+import { WETH98 } from "src/dispute/weth/WETH98.sol";
+
+import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
+
+/// @title DelayedWETH
+/// @notice DelayedWETH is an extension to WETH9 that allows for delayed withdrawals. Accounts must
+///         trigger an unlock function before they can withdraw WETH. Accounts must trigger unlock
+///         by specifying a sub-account and an amount of WETH to unlock. Accounts can trigger the
+///         unlock function at any time, but must wait a delay period before they can withdraw
+///         after the unlock function is triggered. DelayedWETH is designed to be used by the
+///         DisputeGame contracts where unlock will only be triggered after a dispute is resolved.
+///         DelayedWETH is meant to sit behind a proxy contract and has an owner address that can
+///         pull WETH from any account and can recover ETH from the contract itself. Variable and
+///         function naming vaguely follows the vibe of WETH9. Not the prettiest contract in the
+///         world, but it gets the job done.
+contract DelayedWETH is OwnableUpgradeable, WETH98, IDelayedWETH, ISemver {
+    /// @notice Semantic version.
+    /// @custom:semver 0.2.0
+    string public constant version = "0.2.0";
+
+    /// @inheritdoc IDelayedWETH
+    mapping(address => mapping (address => WithdrawalRequest)) public withdrawals;
+
+    /// @notice Withdrawal delay in seconds.
+    uint256 internal immutable DELAY_SECONDS;
+
+    /// @notice Address of the SuperchainConfig contract.
+    SuperchainConfig public config;
+
+    /// @param wait The delay for withdrawals in seconds.
+    constructor(uint256 wait) {
+        DELAY_SECONDS = wait;
+        initialize({ owner: address(0), cfg: SuperchainConfig(address(0)) });
+    }
+
+    /// @notice Initializes the contract.
+    /// @param owner The address of the owner.
+    /// @param cfg Address of the SuperchainConfig contract.
+    function initialize(address owner, SuperchainConfig cfg) public initializer {
+        __Ownable_init();
+        _transferOwnership(owner);
+        config = cfg;
+    }
+
+    /// @inheritdoc IDelayedWETH
+    function delay() external view returns (uint256) {
+        return DELAY_SECONDS;
+    }
+
+    /// @inheritdoc IDelayedWETH
+    function unlock(address guy, uint256 wad) external {
+        require(!config.paused());
+        WithdrawalRequest storage wd = withdrawals[msg.sender][guy];
+        wd.timestamp = block.timestamp;
+        wd.amount += wad;
+    }
```

Comment:
I see that this method is unauthed.

1. What is to stop a DOS by calling this function repeatedly to update the timestamp attached to the withdrawal? 
2. Why is the `wad` amount unlimited, is that safe?

Reply:
The method is sort of authed in the sense that it has a mapping where the first key is `msg.sender`. I agree the code here is a little confusing, basically it's just allowing contracts to do internal accounting while making it explicit inside of this contract. To address your questions:

> What is to stop a DOS by calling this function repeatedly to update the timestamp attached to the withdrawal?

Nothing, the assumption is that the FDG will not call this function an unlimited number of times. If there is a but that could make the FDG call this indefinitely, that would be a reason to consider using a fallback.

> Why is the wad amount unlimited, is that safe?

It's safe because you cannot withdraw more than your balance, even if you try to unlock more than your balance.

> Also, what is to stop any random account from calling this to initiate a withdrawal of someone else's funds?

The `msg.sender` key means that this mapping is on a per-user basis. A third party cannot modify the mapping on behalf of another user.

Reply:
OK, yeah, I see how the auth is moreso done at the level of the first mapping key.

Anyone can write here, but in practice it'll only be meaningful coming from an FDG.

Reply:
Appreciate some comments to that effect tho. :)

Reply:
Comment added